CMakeLists.txt

```diff
@@ -123,6 +81,48 @@ include(Findopm-tests)
 # with the find module
 include ("${project}-prereqs")
 
+# Make sure we are using the same compiler underneath
+# NVCC as for the rest. In the case that NVCC does not support
+# that compiler it will error out. Unfortunately this will only
+# work CMake >= 3.8. We found no way to make FindCUDA.cmake error
+# out. It seems to ignore CMAKE_NVCC_FLAGS and CMAKE. Additionally
+# our way of specifying cuda source files never worked for CMake
+# version < 3.8. Hence we deactivate cuda for these versions.
+# We use "CMAKE_VERSION VERSION_GREATER 3.7.9" instead of
+# CMAKE_VERSION VERSION_GREATER_EQUAL 3.8, because of backwards
+# compatibility to cmake 3.6 and lower.
+if(NOT CMAKE_DISABLE_FIND_PACKAGE_CUDA AND
+    CMAKE_VERSION VERSION_GREATER 3.7.9)
+  if(NOT DEFINED ENV{CUDAHOSTCXX} AND
+      (NOT CMAKE_CUDA_FLAGS OR NOT CMAKE_CUDA_FLAGS MATCHES ".*-ccbin .*"))
+    set(CMAKE_CUDA_FLAGS "-ccbin ${CMAKE_CXX_COMPILER} ${CMAKE_CUDA_FLAGS}")
+  endif()
+  include(CheckLanguage)
+  # Unfortunately check_language(CUDA) will only check whether there is
+  # a nvcc compiler. It will not check whether compilation with the above
+  # option works. Since the OPTIONAL flag of enable_language is currently
+  # ignored CMake will error out below if the options do not work.
+  check_language(CUDA)
+  if(CMAKE_CUDA_COMPILER)
+    enable_language(CUDA OPTIONAL)
+  endif()
+  # While the documentation says that it is deprecated, FindCUDA seems the
+  # only easy way to determine the cublas and cusparse libraries.
+  # Hence we call it unconditionally
+  # The WellContributions kernel uses __shfl_down_sync, which was introduced in CUDA 9.0
+  find_package(CUDA)
+  if(CUDA_VERSION VERSION_LESS "9.0")
```

Comment:
while here, can you please add CUDA_FOUND to the if statement. You now get rather confusing messages if no cuda is found.

Reply:
Done